clients/src/main/java/org/apache/kafka/common/requests/EnvelopeRequest.java

```diff
@@ -0,0 +1,103 @@
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.requests;
+
+import org.apache.kafka.common.message.DefaultPrincipalData;
+import org.apache.kafka.common.message.EnvelopeRequestData;
+import org.apache.kafka.common.message.EnvelopeResponseData;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+import org.apache.kafka.common.security.auth.KafkaPrincipalSerde;
+
+import java.nio.ByteBuffer;
+
+public class EnvelopeRequest extends AbstractRequest {
+
+    public static class Builder extends AbstractRequest.Builder<EnvelopeRequest> {
+
+        private final EnvelopeRequestData data;
+
+        public Builder(ByteBuffer embedData, String clientHostName) {
+            this(embedData, null, clientHostName);
+        }
+
+        public Builder(ByteBuffer embedData,
+                       ByteBuffer serializedPrincipal,
+                       String clientHostName) {
+            super(ApiKeys.ENVELOPE);
+            this.data = new EnvelopeRequestData()
+                            .setRequestData(embedData)
+                            .setRequestPrincipal(serializedPrincipal)
+                            .setClientHostName(clientHostName);
+        }
+
+        @Override
+        public EnvelopeRequest build(short version) {
+            return new EnvelopeRequest(data, version);
+        }
+
+        @Override
+        public String toString() {
+            return data.toString();
+        }
+    }
+
+    private final EnvelopeRequestData data;
+
+    public EnvelopeRequest(EnvelopeRequestData data, short version) {
+        super(ApiKeys.ENVELOPE, version);
+        this.data = data;
+    }
+
+    public EnvelopeRequest(Struct struct, short version) {
+        super(ApiKeys.ENVELOPE, version);
+        this.data = new EnvelopeRequestData(struct, version);
+    }
+
+    public ByteBuffer requestData() {
+        return data.requestData();
+    }
+
+    public String clientHostName() {
+        return data.clientHostName();
+    }
+
+    public KafkaPrincipal requestPrincipal(KafkaPrincipalSerde principalSerde) {
```

Comment:
nit: I think it might be better to pull this out of the request class. The direction we're moving is toward dumber request/response classes. Eventually `EnvelopeRequest` will go away and we'll just use `EnvelopeRequestData`.